crates/bevy_ecs/src/error/handler.rs

```diff
@@ -77,51 +75,75 @@ impl ErrorContext {
-///
-/// ```rust
-/// use bevy_ecs::error::{default_error_handler, GLOBAL_ERROR_HANDLER, BevyError, ErrorContext, panic};
-///
-/// fn handle_errors(error: BevyError, ctx: ErrorContext) {
-///    let error_handler = default_error_handler();
-///    error_handler(error, ctx);        
-/// }
-/// ```
-///
-/// # Warning
-///
-/// As this can *never* be overwritten, library code should never set this value.
-#[cfg(feature = "configurable_error_handler")]
-pub static GLOBAL_ERROR_HANDLER: OnceLock<fn(BevyError, ErrorContext)> = OnceLock::new();
+mod global_error_handler {
+    use super::{BevyError, ErrorContext};
+    use bevy_platform_support::sync::atomic::{
+        AtomicPtr,
+        Ordering::{AcqRel, Acquire, Relaxed},
+    };
+
+    // SAFETY: Only accessible from within this module.
+    static HANDLER_ADDRESS: AtomicPtr<()> = AtomicPtr::new(core::ptr::null_mut());
+
+    /// Set the global error handler.
+    ///
+    /// If used, this should be called [before] any uses of [`default_error_handler`],
+    /// generally inside your `main` function before initializing the app.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_ecs::error::{set_global_default_error_handler, warn};
+    /// set_global_default_error_handler(warn);
+    /// // initialize Bevy App here
+    /// ```
+    ///
+    /// To use this error handler in your app for custom error handling logic:
+    ///
+    /// ```rust
+    /// use bevy_ecs::error::{default_error_handler, BevyError, ErrorContext, panic};
+    ///
+    /// fn handle_errors(error: BevyError, ctx: ErrorContext) {
+    ///    let error_handler = default_error_handler();
+    ///    error_handler(error, ctx);        
+    /// }
+    /// ```
+    ///
+    /// # Warning
+    ///
+    /// As this can *never* be overwritten, library code should never set this value.
+    ///
+    /// [before]: https://doc.rust-lang.org/nightly/core/sync/atomic/index.html#memory-model-for-atomic-accesses
+    /// [`default_error_handler`]: super::default_error_handler
+    pub fn set_global_default_error_handler(handler: fn(BevyError, ErrorContext)) {
+        if HANDLER_ADDRESS
+            .compare_exchange(core::ptr::null_mut(), handler as *mut (), AcqRel, Acquire)
+            .is_err()
+        {
+            panic!("Global error handler set multiple times");
+        }
+    }
+
+    #[inline]
+    pub(super) fn get() -> Option<fn(BevyError, ErrorContext)> {
+        // The error handler must have been already set from the perspective of this thread,
+        // otherwise we will panic. It will never be updated after this point.
+        // We therefore only need a relaxed load.
+        let ptr = HANDLER_ADDRESS.load(Relaxed);
+        // SAFETY: We only ever store null or a valid handler function.
+        // `null` is guaranteed to be represent `None`:
+        // https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html#representation
+        unsafe { core::mem::transmute(ptr) }
```

Comment:
The rust docs have this to say on the topic of raw-pointer to function-pointer transmutation:

> Note that all of this is not portable to platforms where function pointers and data pointers have different sizes.

Is this something we are concerned about? If not, it would probably be good to mention it in or around the safety comment there.

Reply:
There's a crate called `atomic_fn` that does gross int-to-pointer casting but which seems to solve the size issue. I don't think we need to use it, merely mentioning it for posterity in case this becomes an issue later.